Avoid unnecessary typecast in java_bytecode_convert_methodt::variable

romainbrenguier opened the pull request:
~Based on https://github.com/diffblue/cbmc/pull/4870~
When the variable is a reference and the type_char is 'a' then typecast is not required.

This change is to prevent unnecessary typecast which can lead to byte_extract operations later down the line and are more difficult to deal with.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

romainbrenguier committed:
Remove useless check for typecast

We no longer need this cast, as `variable` now returns the expression
with its correct type.

romainbrenguier committed:
Unit test for convert_store

This function was not directly tested.

romainbrenguier committed:
Unit test java_bytecode_convert_methodt::variable

This method was not directly tested.
This illustrates the fact that we do not typecast from reference of
array[int] to reference of void.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 53072bff0a7f20277685ee10065d78ffc53e567d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/118054161

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c9b8ad7016b600e995e3ce1675773f73918a65a9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/118063592

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 5735d32d3de1441cbfae827e2c5cb140b6aca336).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/118063322
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

thk123 commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -202,7 +202,12 @@ exprt java_bytecode_convert_methodt::variable(
     if(do_cast == CAST_AS_NEEDED)
+    {
+      /// No need for a cast if result type and `t` are reference types
+      if(type_char == 'a' && can_cast_type<java_reference_typet>(result.type()))
```
:question: Would a more sensible fix be to make `'a'` give `reference(java.lang.Object)`? so the types actually match?

thk123 replied:
:bulb: Might be clearer to use `t` here rather than `type_char`

romainbrenguier replied:
'a' can be any reference no just to an Object I think, so it could maybe reference primitive types, or arrays of primitive types

thk123 replied:
I'm not sure I follow:
 - a reference to a primitive is done via boxing 
 - an array of primitives is an `instanceof java.lang.Object`

So seems like both should be thought of and can be assigned into a `java_reference_type(java.lang.Object)`

romainbrenguier replied:
I don't know if this is correct java but in Jbmc we have reference to primitive types (unboxed) and void in particular.

peterschrammel replied:
Do you have an example how these are produced?

romainbrenguier replied:
https://github.com/diffblue/cbmc/blob/develop/jbmc/src/java_bytecode/java_bytecode_convert_class.cpp#L824
https://github.com/diffblue/cbmc/blob/develop/jbmc/src/java_bytecode/java_bytecode_convert_method.cpp#L1259
https://github.com/diffblue/cbmc/blob/develop/jbmc/src/java_bytecode/java_bytecode_convert_method.cpp#L2019
https://github.com/diffblue/cbmc/blob/develop/jbmc/src/java_bytecode/java_bytecode_convert_method.cpp#L2736
https://github.com/diffblue/cbmc/blob/develop/jbmc/src/java_bytecode/java_bytecode_convert_method.cpp#L2758
https://github.com/diffblue/cbmc/blob/develop/jbmc/src/java_bytecode/java_entry_point.cpp#L697

romainbrenguier replied:
> Might be clearer to use t here rather than type_char

But how? `t != java_type_from_char('a')`? I don't think that's clearer.

thk123 commented on jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp:
```diff
@@ -308,6 +308,34 @@ class java_bytecode_convert_method_unit_testt
+    const exprt &arg,
+    char type_char,
+    size_t address,
+    bool cast_as_needed)
```
:pick: Why not just take `java_bytecode_convert_methodt::variable_cast_argumentt`

romainbrenguier replied:
this type is protected

thk123 commented on jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp:
```diff
@@ -308,6 +308,34 @@ class java_bytecode_convert_method_unit_testt
+                     : java_bytecode_convert_methodt::NO_CAST);
+  }
+
+  static void add_variable(
```
:bulb: Since isn't just a wrapper around `java_bytecode_convert_methodt` might be worth documenting

romainbrenguier replied:
With the refactoring that removes the extra argument, it is now just a trivial wrapper

thk123 commented on jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp:
```diff
@@ -518,3 +546,54 @@ SCENARIO(
+    const symbol_exprt int_array{"int_array", int_array_type};
+    const std::size_t variable_index = 0;
+    java_bytecode_convert_method_unit_testt::add_variable(
+      converter, variable_index, int_array, 0, 1, false, {});
```
:pick: Since lots of similar typed parameters you might like to pull these out into named constants before:

```C++
const std::size_t start_pc =  0;
...
const bool is_parameter = false;
java_bytecode_convert_method_unit_testt::add_variable(
  converter, variable_index, int_array, start_pc, length, is_paramerter, {});
```

:point_left:

thk123 commented on jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp:
```diff
@@ -518,3 +546,54 @@ SCENARIO(
+    const std::size_t address = 0;
+    WHEN("variable is called as it would be for an `aload` instruction")
+    {
+      const exprt index_expr = from_integer(variable_index, java_int_type());
```
:pick: `constant_exprt` as with a name like `index_expr` I was expecting an `index_exprt`...

thk123 commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -2911,7 +2912,7 @@ code_blockt java_bytecode_convert_methodt::convert_store(
     var_name);
 
-  block.add(code_assignt{var, toassign}, location);
+  block.add(code_assignt{var, op[0]}, location);
```
:question: I'd expect to see a unit test / change to existing unit test for this change no?

thk123 commented on jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp:
```diff
@@ -518,3 +546,54 @@ SCENARIO(
+  const class_hierarchyt class_hierarchy;
+  java_bytecode_convert_methodt converter{symbol_table,
+                                          null_message_handler,
+                                          10,
```
:point_left:

thk123 commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -202,7 +202,12 @@ exprt java_bytecode_convert_methodt::variable(
+      /// No need for a cast if result type and `t` are reference types
+      if(type_char == 'a' && can_cast_type<java_reference_typet>(result.type()))
+        return result;
       result = typecast_exprt::conditional_cast(result, t);
```
:bulb: Don't suppose you fancy giving `t` a better name (`access_type`?)

thk123 commented on jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp:
```diff
@@ -518,3 +546,54 @@ SCENARIO(
+    }
+    WHEN("There is no variable at the given index")
+    {
+      const exprt index_expr =
```
:question: The variable doesn't exist, so it just makes one up??

romainbrenguier replied:
yes, it is used for anonymous variables, I think this happens for function arguments. If you see "anonlocal" somewhere, it should come from that.

thk123 commented on jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp:
```diff
@@ -518,3 +546,54 @@ SCENARIO(
+        REQUIRE(result == int_array);
+      }
+    }
+    WHEN("There is no variable at the given index")
```
:bulb: Ideally we'd test a non-array reference type, and a primitive type (not sure what the latter should do, maybe throw?)

thk123 reviewed (commented):
Couple of questions for you - I think I get it, but to be sure I might try updating the docs on `variable` to check I understand its purpose...

romainbrenguier committed:
Get rid of t variable in java bytecode convert variable

This is only used once and the name does not give any information.
Also refactor a bit the else branch to simplify the function.

romainbrenguier committed:
Let caller of variable do typecast if needed

The variable method is mostly called with NO_CAST so it is simpler to
let the caller do the typecast in the rare case they need it.

romainbrenguier committed:
Extract a convert_load function form convert_instructions

This is similar to how other instructions are handled, makes the
convert_instructions method shorter and gives a natural place where to
document the conversion of the load instruction.

romainbrenguier committed:
Make convert_load match its documentation

Rather than typecasting as needed, we check that the type corresponds to
the instruction.
In particular this removes the casting of reference to void reference in
the case of `a`.

romainbrenguier commented:
@thk123 after discussing with @peterschrammel I decided to refactor a bit by defining a `convert_load` method which should make things clearer.

romainbrenguier committed:
Unit test convert_load

Test the behaviour of the method on different type of inputs.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 54a45f5bd9a6257e0c62247230787e7dfe994ebc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/118390989

smowton commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -164,13 +164,11 @@ symbol_exprt java_bytecode_convert_methodt::tmp_variable(
 /// from a bytecode at address `address` a value of type `type_char` stored in
 /// the JVM's slot `arg`.
```
Commit message: convertion -> conversion

smowton commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -164,13 +164,11 @@ symbol_exprt java_bytecode_convert_methodt::tmp_variable(
 /// the JVM's slot `arg`.
 /// \param arg: The local variable slot
-/// \param type_char: The type of the value stored in the slot pointed by `arg`
+/// \param type_char: The type of the value stored in the slot pointed by `arg`,
```
pointed *to* by

smowton commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -164,13 +164,11 @@ symbol_exprt java_bytecode_convert_methodt::tmp_variable(
 /// \param arg: The local variable slot
-/// \param type_char: The type of the value stored in the slot pointed by `arg`
+/// \param type_char: The type of the value stored in the slot pointed by `arg`,
+///   this is only used in case a new unnamed local variable is created
```
in the case where a...

smowton commented on jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp:
```diff
@@ -646,3 +655,144 @@ SCENARIO(
+      }
+    }
+  }
+  GIVEN("An long variable")
```
A long (similar below)

smowton commented on jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp:
```diff
@@ -308,6 +308,28 @@ class java_bytecode_convert_method_unit_testt
   {
     return converter.convert_astore(statement, op, location);
```
Commit message: "This illustrate the" -> "This illustrates the"

smowton reviewed (approved):
Just English stuff. Suggest writing in the PR description *why* this was done -- I expect to make trace processing easier?

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e8f0110730bcdac3f1dac9d1dbd40988edbff428).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/118456920

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 611daa1e9cf191885429139ba2fa8f13808d5a36).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/118469719

thk123 commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -164,13 +164,12 @@ symbol_exprt java_bytecode_convert_methodt::tmp_variable(
 /// \param address: Bytecode address used to find a variable that the LVT
 ///   declares to be live and living in the slot pointed by `arg` for this
 ///   bytecode
-/// \param do_cast: Indicates whether we should return the original symbol_exprt
```
:pick: This ends up in the wrong commit

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: da172f73691ef7cec1dcc9ab80867af48c2d8fd3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/118586502

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7dbb9a123c5117b2efde7524269970776b4a149e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/118656853

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 14cd977938fafb8cbec7c245725a300a9034b625).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/118671364

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 1c93adbab2da50d6f36bf53e882b263f7ed834d0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/118721038
